### ISSUE TITLE
left and right audio channels are swapped (in Space Invaders?)

### ISSUE THREAD

**Issue by mrehkopf** [redacted173/sd2snes#13]
There seems to be another problem a problem with space invaders:
Mr_Boo_Berry: Space Invaders - Stereo is reversed from what it should be (generally, this is the best game to test)
Ikari_01: hm, I just tested on sd2snes (sgb00 however) and the channels aren't reversed
yeah, they are swapped on sgb03 ^^

_Originally posted by @Conn79 in https://github.com/redacted173/sd2snes/issues/12#issuecomment-658256358_

**Comment by mrehkopf** [comment]
It would appear that the issue surfaced in sgb02 where the APU audio data vector was refactored into separate sub-vectors per channel. At first glance I can't see any issues with it though (no obvious confusion of the left/right mapping)...
Not sure if only Space Invaders is affected but it's good for testing (in Super Gameboy Version mode, move to the left screen edge and shoot, then move to the right screen edge and shoot).

**Comment by redacted173** [comment]
I'm having difficulty reproducing the stereo reverse problem in sgb03 vs original hardware, but it looks like some clone systems (e.g. SuperNT) may have it swapped by default.  The way I tested this is:

- started Space Invaders (SGB Enhanced) game
- selected "Super GameBoy" version (not Arcade)
- selected "Color"

In the game I then shoot on the left side of the screen and then on the right to confirm the sound is on the left and right channel, respectively.  Listed below is whether it is "correct" or "reverse".

SNES 1-chip + original SGB2: correct
SNES 1-chip + SD2SNES sgb00: reverse
SNES 1-chip + SD2SNES sgb01: correct
SNES 1-chip + SD2SNES sgb03: correct
SNES 1-chip + SD2SNES recent unreleased changes: correct

So it looks like a bug was fixed a while back in sgb01.  But when using SuperNT:

SuperNT + original SGB2 + "Swap Left & Right" unchecked: reverse
SuperNT + SD2SNES sgb03 + "Swap Left & Right" unchecked: reverse

Are people testing this on SuperNT or maybe another clone?  If so, looks like the "Swap Left & Right" option needs to be checked to get the audio channels correct on some or all versions of its firmware.  Unfortunately, when that option is selected it also reverses the SNES APU output where that is no longer correct as tested in the "SNES Test Program" Audio section.

I've also noticed a lot less flickering on the first and last scanline (typically overscan region on CRTs) with SuperNT.  On a 1-chip the top line tends to flicker a lot when it has a full background image for both the original SGB2 and SD2SNES.  The bottom line also tends to flicker a bit, too.  See Pokemon Gold and Donkey Kong for examples of this.

Assuming I didn't mess anything up in the testing the SuperNT differences vs original SNES seem relatively minor.  Although the audio being swapped is unfortunate.  It would be good for people to test vs as much original hardware as possible, but I understand that not everyone has access to that.  Any testing and feedback is appreciated.

**Comment by mrehkopf** [comment]
I can confirm that I seem to have the channels swapped somewhere. *sigh* Nowhere obvious so it must be in the SNES RGB cable or the OSSC (stereo jack output) or the phono-cinch cable itself. Panning on my audio mixer gets translated correctly to headphones.
Thanks for the detailed testing, there seems to be a lot to be unearthed about potential pitfalls in people's audio setups :)

**Comment by redacted173** [comment]
We've seen in the past that my audio setup has its own share of bugs.  I had to run the tests above multiple times and I still am not 100% sure of the findings.  At least the original SGB2 and SD2SNES sgb03 are consistent in the different setups which is the key result.

**Comment by GreatHierophant** [comment]
Okay, I did more testing with Space Invaders and found that I was wrong.
The stereo is not reversed with the sd2snes.  I checked an original SNES with the sd2snes and my Super Game Boy 2 and both behave correctly with Space Invaders : When you are on the left side of the screen, you hear the shot only from the left speaker, when you are in the middle you hear the shot from both speakers and when you are on the right side of the screen you hear the shot only from the right speaker.
Nor is the stereo reversed with the Super Nt.  Kevtris fixed that in a firmware update, but I had my RCA cables swapped as connected to the DAC.

This bug report should be closed, sorry for the inconvenience.
